### PR TITLE
Fix crash in element access signature help

### PR DESCRIPTION
**Pull request by rchande**
We previously assumed that conditional element access would never be followed by a member or conditional access.

Fixes https://github.com/dotnet/roslyn/issues/20507

Tag @dotnet/roslyn-ide

### CODE REVIEW

**Comment by rchande** [comment]
retest windows_release_unit32_prtest

**Comment by rchande** [comment]
Unit 32 failure was unrelated, as were VSI failures.

**Comment by rchande** [comment]
Tag @MattGertz for approval

**Customer scenario**
Customer uses conditional indexing and dots into the resulting object. If they start editing the indexer, signature help will crash.

**Bugs this fixes:** #20507
**Workarounds, if any**: None
**Risk**: Low: fixes an invalid cast by explicitly walking the tree for the right type of node.
**Performance impact**: Low: walks the tree slightly more than we used to in order to provide the correct behavior
**Is this a regression from a previous update?**: No, we've had this since we added signature help for conditional element access
**Root cause analysis:**: The code was not tested with complicated trees
**How was the bug found?**: VSfeedback

File: src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs [hunk 1]
```diff
@@ -372,7 +372,18 @@ internal static bool TryGetSyntax(SyntaxNode root, int position, ISyntaxFactsSer
             {
                 if (CommonSignatureHelpUtilities.TryGetSyntax(root, position, syntaxFacts, triggerReason, IsTriggerToken, IsArgumentListToken, cancellationToken, out ElementBindingExpressionSyntax elementBindingExpression))
                 {
-                    identifier = ((ConditionalAccessExpressionSyntax)elementBindingExpression.Parent).Expression;
+                    // Find the first conditional access expression that starts left of our open bracket
+                    var conditionalAccess = elementBindingExpression.FirstAncestorOrSelf<ConditionalAccessExpressionSyntax>(
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: this seems wrong.  you could have:

```a?[....actualStuff[```

Won't this find that outer conditional access?
Won't this

Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: Also, is this just suprressing signature help?  don't we want sighelp to work here?

Comment (rchande) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: I think this is correct `x?[y[$$]?Count]?.Count;` shows sighelp for y's indexer.

Comment (rchande) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: Likewise sighelp between [ and y shows info for x's indexer.

Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: ok.  got it.  you have an element binding expressoin.  so it should *always* have a contaiing ConditionalAccessExpression right?

File: src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs [hunk 2]
```diff
@@ -372,7 +372,18 @@ internal static bool TryGetSyntax(SyntaxNode root, int position, ISyntaxFactsSer
...
+                        c => c.SpanStart < elementBindingExpression.SpanStart);
+
+                    if (conditionalAccess == null)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: how is this ever null?

File: src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs (hunk 1 above)
Comment (rchande) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: That's right.

File: src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs (hunk 2 above)
Comment (rchande) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: Yeah, this isn't necessary.

File: src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/SignatureHelp/ElementAccessExpressionSignatureHelpProvider.cs]: we have: GetParentConditionalAccessExpression

**Comment by cston** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Pull request merged**